### DISCUSSION TITLE
Maui xaml x:static好像不支持从继承类的父类中读取const 常量？？

### DISCUSSION THREAD

**Discussion by sfmmwl** [BYJRK/DotNet-Discussions#110] [其他框架]
maui xaml的x:static 好像不支持从继承类的父类里获取常量，以前wpf是可以的。
比如有两个类class A和 class B:A，
A有public const string BaseString="string A"
B有public const string ClassString="string B"

c#代码和maui xaml x:static调用A.BaseString和B.ClassString都没有问题。
c#代码里调用B.BaseString没问题，但在maui xaml x:static调用B.BaseString运行时就出错。错误信息大约就是无法找到静态成员B.BaseString。

示例项目源码参见
https://github.com/sfmmwl/Net7MauiTest

.net 版本7.0.81
开发环境 vs2022 17.6.2

发过官方issue，但可能是英语水平问题描述不清还是其它原因，只收到一条认为没有问题的回复。
哪位有兴趣帮忙一起看看分析一下？ 如果是我的vs的设置问题，帮忙告知一下。如果确定是maui的问题，也请帮忙反馈给官方团队。

**Comment by KiyanYang** [comment]
**分析**：

查看 Maui StaticExtension [源码](https://github.com/dotnet/maui/blob/7bc2e2537395de0c76990647b880e771b67d0061/src/Controls/src/Xaml/MarkupExtensions/StaticExtension.cs#L33)，获取字段的代码如下：

`var finfo = type.GetRuntimeFields().FirstOrDefault(fi => fi.Name == membername && fi.IsStatic);`

其中使用 `GetRuntimeFields()` 方法来获取字段，通过[源代码](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Reflection/RuntimeReflectionExtensions.cs,e1e8858f5ed384cf)或文档可知，该方法获取到的字段不包括基类静态字段。

**解决方法**：

- 改为 `<Label Text="{x:Static main:BaseClass.BaseClassString}" />`

- 创建自己的 XAML 标记扩展（简单修改 MAUI StaticExtension）：

  ```cs
  using System.Reflection;

  namespace Net7MauiTest;

  [ContentProperty(nameof(Member))]
  public class MyStaticExtension : IMarkupExtension
  {
      public string Member { get; set; }

      public object ProvideValue(IServiceProvider serviceProvider)
      {
          if (serviceProvider == null)
              throw new ArgumentNullException(nameof(serviceProvider));
          if (!(serviceProvider.GetService(typeof(IXamlTypeResolver)) is IXamlTypeResolver typeResolver))
              throw new ArgumentException("No IXamlTypeResolver in IServiceProvider");
          if (string.IsNullOrEmpty(Member) || Member.IndexOf(".", StringComparison.Ordinal) == -1)
              throw new XamlParseException("Syntax for x:Static is [Member=][prefix:]typeName.staticMemberName");

          var dotIdx = Member.LastIndexOf('.');
          var typename = Member.Substring(0, dotIdx);
          var membername = Member.Substring(dotIdx + 1);

          var type = typeResolver.Resolve(typename, serviceProvider);

          var everything = BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Static | BindingFlags.Instance | BindingFlags.FlattenHierarchy;

          var pinfo = type.GetProperties(everything).FirstOrDefault(pi => pi.Name == membername && pi.GetMethod.IsStatic);
          if (pinfo != null)
              return pinfo.GetMethod.Invoke(null, new object[] { });

          var finfo = type.GetFields(everything).FirstOrDefault(fi => fi.Name == membername && fi.IsStatic);
          if (finfo != null)
              return finfo.GetValue(null);

          throw new XamlParseException($"No static member found for {Member}");
      }
  }

  ```

  使用 `<Label Text="{main:MyStatic main:TestClass.BaseClassString}" />`

**Reply by sfmmwl** [reply]
非常感谢帮忙分析和提供建议。
写xaml时vs的代码智能补全也能跳出父类里的常量供选择，然而实际上却不能用。真是个奇怪的设定。

有兴趣的话，能否帮忙给官方提个需求？我自己的英文水平有限估计又会写别人看不懂的东西。
从你的参考代码来看感觉不是很难实现，毕竟wpf xaml里是允许这样调用，到maui xaml却取消了就很迷。

**Reply by sfmmwl** [reply]
我把maui源码和你的代码结合了一下，先按maui GetRuntimeProperties，没结果则继续GetProperties。flag取消nonpublic和instance。代码win平台和android平台测试有效，再次对maui的做法感到迷惑。

不过实际在xaml里使用这个自定义标记时，写后续常量时vs的代码智能补全似乎就失效了，大概不能识别这个标记的缘故，也是个小遗憾。
